File: rout_new/MakeConvolution.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include "rout.h"
/**************************************/
/* MakeConvolution                    */ 
/**************************************/
void MakeConvolution(int number_of_cells,
                     int skip,
		     int ndays,
		     int **CATCHMENT,
		     ARC **BASIN,
		     double *BASEFLOW,
		     double *RUNOFF,
		     double *FLOW,
		     float **UH_S,
		     LIST *STATION,
		     int nr,
		     float xllcorner,
		     float yllcorner,
		     float size,
		     char *inpath,
		     char *outpath,
		     char *workpath,
		     int decimal_places,
		     TIME *DATE,
		     float *factor_sum,
		     int first_year,
		     int first_month,
		     int irr_rout,
		     int missing)
{
  FILE *fp;
  int i,j,n,ii,jj;
  int row,col;
  double factor;
  char infile[500];
  char LATLON[50];
  char fmtstr[17];
  char leftover[MAXSTRING];
  double area,area_sum;
  double radius;
  float k_const;
  float dummy;
  float lat,lon;

  k_const=1.0;
  area_sum=(*factor_sum)=0.0;

  for(i=1;i<=ndays;i++) 
    FLOW[i] = 0.0;
  
  for(n=1;n<=number_of_cells;n++) { //the gridcell loop
    for(i=1;i<=ndays;i++) {
      RUNOFF[i]=0.0;
      BASEFLOW[i]=0.0;
    }
    
    ii = CATCHMENT[n][0];  //row from bottom 
    jj = CATCHMENT[n][1]; //col from left
    lat=yllcorner + ii*size - size/2.0;
    lon=xllcorner + jj*size - size/2.0;
    //printf("Makeconv cellnumber%d row%d col%d %f %f\n",n,ii,jj,lat,lon);
    
    /* Give area of box in square kilometers */
    radius = (double)EARTHRADIUS;
    area = radius*radius*fabs(size)*PI/180*            
      fabs(sin((lat-size/2.0)*PI/180)-     
	   sin((lat+size/2.0)*PI/180));
    //if(CATCHMENT[n][2]!=2) 
    area_sum += area;
    printf("area_sum=%f area=%f routed %d\n",area_sum,area,BASIN[ii][jj].routed);

    /* Find conversion factor for mm/day to m3/s
       and multiply by cell fraction */
    factor = BASIN[ii][jj].fraction*area/86.4;  
    (*factor_sum) += factor;

    if(BASIN[ii][jj].routed==0 || 
       (irr_rout==1 && CATCHMENT[n][2]==0)) {
      /* Make vic filename */
      strcpy(infile,inpath);
      sprintf(fmtstr,"%%.%if_%%.%if",
	      decimal_places,decimal_places);
      sprintf(LATLON,fmtstr,lat,lon);
      strcat(infile,LATLON);
      printf("File %d of %d: %s\n",n,number_of_cells,infile);

      if((fp = fopen(infile,"r"))==NULL) { 
	printf("Cannot open file, inserting zeroes...%s \n",infile);
	for(i=1;i<=ndays;i++) {
	  DATE[i].year=0;
	  DATE[i].month=0;
	  DATE[i].day=0;
	  RUNOFF[i]=0.;
	  BASEFLOW[i]=0.;
	}
	//printf("Cannot open file, exiting...%s \n",infile);
	//exit(0);
      }
      else {
	/* Read VIC model output: 
	   <year> <month> <day> <p> <et> <runoff> <baseflow>*/
	for(i=1;i<=skip;i++) 
	  fgets(leftover,MAXSTRING,fp); 
	for(i=1;i<=ndays;i++) {
	  fscanf(fp,"%d %d %d %lf %lf ",
		 &DATE[i].year,&DATE[i].month,&DATE[i].day,
		 &RUNOFF[i],&BASEFLOW[i]);
	  /* Check to be sure dates in VIC file start at same time 
	     specified in input file */
	  if(i == 1) {
	    if(DATE[i].year!=first_year || DATE[i].month!=first_month) {
	      printf("VIC output file does not match specified\n");
	      printf("period in input file.%d %d %d %d\n",DATE[i].year,first_year,DATE[i].month,first_month);
	      exit(2);
	    }
	  }
	}
	fclose(fp);
      }
    }
    else {
      if(BASIN[ii][jj].routed==2) {
	/* Flow already routed at this location, read from file */
        printf("MakeConvolution, flow already routed, read from file\n");
	strcpy(infile,workpath);
	sprintf(fmtstr,"streamflow_%%.%if_%%.%if",decimal_places,decimal_places);
	sprintf(LATLON,fmtstr,lat,lon);
	strcat(infile,LATLON);
	printf("File %d of %d: %s\n",n,number_of_cells,infile);
	
	if((fp = fopen(infile,"r"))==NULL) { 
	  printf("Cannot open file %s \n",infile);
	  exit(0);
	}
	else {
	  /* Read routed output (m3/s): 
	     <year> <month> <day> <total runoff> */
	  for(i=1;i<=ndays;i++) {
	    fscanf(fp,"%d %d %d %lf ",
		   &DATE[i].year,&DATE[i].month,&DATE[i].day,
		   &RUNOFF[i]);
	    BASEFLOW[i]=0;
	    /* Check to be sure dates in routed file start at same time 
	       as specified */
	    if(i == 1) {
	      if(DATE[i].year!=first_year || DATE[i].month!=first_month) {
		printf("Routed output file does not match specified\n");
		printf("period.\n");
		exit(2);
	      }
	    }
	  }
	  fclose(fp);
	}	 
      }
    }

    if(BASIN[ii][jj].routed!=1) {
      if(BASIN[ii][jj].routed==0 || CATCHMENT[n][2]==0) {
	for(i=1;i<=ndays;i++) {
	  RUNOFF[i] = RUNOFF[i]*factor;
	  BASEFLOW[i] = BASEFLOW[i]*factor;
	}
      }
      for(i=1;i<=ndays;i++) {
	for(j=1;j<KE+UH_DAY;j++) {
	  if((i-j+1)>=1) {
	    FLOW[i] += UH_S[n][j]*(BASEFLOW[i-j+1]+RUNOFF[i-j+1]);
	  }
	}
      }
    }

    row=CATCHMENT[n][0];
    col=CATCHMENT[n][1];
    if(row==STATION[nr].row && col==STATION[nr].col) 
       BASIN[row][col].routed=2;
    else {
      BASIN[row][col].direction=missing;
      BASIN[row][col].routed=1;
    }
  } /* End gridcell loop */
}

File: src/make_convolution.f

      SUBROUTINE MAKE_CONVOLUTION
     & (NCOL, NROW, NOB, PMAX, DAYS, CATCHIJ, 
     &  BASE, RUNO, FLOW, KE, UH_DAY, UH_S, FRACTION, FACTOR_SUM,
     &  XC, YC, SIZE, DPREC, INPATH,ICOL,NDAY,IDAY,IMONTH,IYEAR,
     &  MO, YR, NYR)

      IMPLICIT NONE

c     RCS ID STRING
      CHARACTER*50 RCSID
      DATA RCSID/"$Id: make_convolution.f,v 1.1 2005/04/07 05:07:28 vicadmin Exp $"/

      INTEGER     N, I, J, DAYS, NDAY, II, JJ  
      INTEGER     NCOL,NROW,ICOL,NOB,PMAX,KE,UH_DAY
      INTEGER     CATCHIJ(PMAX,2)
      INTEGER     NYR
      REAL        UH_S(PMAX,KE+UH_DAY-1)
      REAL        BASE(DAYS), RUNO(DAYS), FLOW(DAYS) 
      REAL        FRACTION(NCOL,NROW)

      REAL        PI, RERD, FACTOR, FACTOR_SUM

      LOGICAL TORF

      PARAMETER   (RERD  = 6371229.0)    !radius of earth in meters

      CHARACTER*20 LOC
      CHARACTER*72 INPATH

      INTEGER DPREC, CLEN

      REAL        JLOC, ILOC
      REAL        XC, YC, SIZE
      REAL        AREA, AREA_SUM

      REAL        STORAGE, K_CONST
      REAL        DUM1,DUM2
  
      INTEGER     IDAY(DAYS), IMONTH(DAYS), IYEAR(DAYS)
      INTEGER MO(12*NYR),YR(12*NYR)
      INTEGER MISS_NUM
C     MISS_NUM is the number of grid cell output files not found

      MISS_NUM=0

C     *** 0 <= K_CONST = 1.0 
C *** K_CONST smaller 1.0 makes it a simple linear storage
    
      K_CONST = 1.0

      PI = ATAN(1.0) * 4.0

      AREA_SUM   = 0.0
      FACTOR_SUM = 0.0

      DO I = 1,NDAY
         FLOW(I) = 0.0
      END DO

      DO N = 1,NOB       !is this the gridcell loop?
         STORAGE = 0.0
         DO I = 1,NDAY
            RUNO(I) = 0.0
            BASE(I) = 0.0
         END DO
         II = CATCHIJ(N,1)
         JJ = CATCHIJ(N,2)
         
c     the grid has been flipped left to right
c     find the revised cooordinates

         ILOC=XC + (ICOL-II)*SIZE + SIZE/2.0
         JLOC=YC + JJ*SIZE - SIZE/2.0

C        CONVERSIONFACTOR for mm/day to ft**3/sec         <--?????


         AREA =  RERD**2*ABS(SIZE)*PI/180*             !give area of box in 
     &        ABS(SIN((JLOC-SIZE/2.0)*PI/180)-         !square meters
     $        SIN((JLOC+SIZE/2.0)*PI/180))

         
         AREA_SUM = AREA_SUM + AREA

c         WRITE(*,*) N, ILOC, JLOC

         FACTOR = FRACTION(II,JJ)*35.315*AREA/(86400.0*1000.0)  !convert to sq.mi.
                                                                !&mult. by cell fract
         FACTOR_SUM = FACTOR_SUM + FACTOR

         
         call create_vic_names(jloc,iloc,loc,clen,dprec)

c     print*, INPATH(1:INDEX(INPATH,' ')-1)//LOC(1:CLEN)

         INQUIRE(FILE=INPATH(1:(INDEX(INPATH,' ')-1))//LOC(1:CLEN),
     $                     EXIST=TORF)

         if(torf)then
           OPEN(20,FILE=INPATH(1:(INDEX(INPATH,' ')-1))//
     $        LOC(1:CLEN),
     $        STATUS='OLD',ERR=9001)
           print*, N, ' of',NOB,": ",
     &       INPATH(1:(INDEX(INPATH,' ')-1))//LOC(1:CLEN)
c     read VIC model output: <year> <month> <day> <p> <et> <runoff> <baseflow>
           DO I = 1,NDAY
             READ(20,*,END=9001,ERR=9001) IYEAR(I), IMONTH(I), IDAY(I),
     &         DUM1, DUM2, RUNO(I), BASE(I)
c     check to be sure dates in VIC file start at same time specified
c     in input file
             if(I.eq.1) then
               if(IYEAR(I).ne.YR(1) . or. IMONTH(I).ne.MO(1)) then
                  print*, 'VIC output file does not match specified '
                  print*, 'period in input file.'
                  stop
               endif
             endif
           END DO
	 else
           print*, INPATH(1:(INDEX(INPATH,' ')-1))//LOC(1:CLEN),
     &             ' NOT FOUND, INSERTING ZEROS'
           miss_num = miss_num+1
	   do i=1,nday
             IYEAR(I)=9999
             IMONTH(I)=99
             IDAY(I)=99
	     runo(i)=0
	     base(i)=0
	   end do
	 endif

         DO I = 1,NDAY
            RUNO(I) = RUNO(I) * FACTOR
            BASE(I) = BASE(I) * FACTOR
         END DO
         DO I = 1,NDAY
            DO J = 1,KE+UH_DAY-1
               IF ((I-J+1) .GE. 1) THEN
                  FLOW(I) = FLOW(I)+UH_S(N,J)*(BASE(I-J+1)+RUNO(I-J+1))
               END IF
            END DO
         END DO
         CLOSE(20)
      END DO
      if(MISS_NUM>0) then
        print*, MISS_NUM, ' files not found, zero runoff/baseflow used'
      end if

      RETURN
 9001 WRITE(*,*) 'Error reading time-series data, ',
     $     'insufficient data or missing input file',
     $     INPATH(1:INDEX(INPATH,' ')-1)//LOC(1:CLEN)
      END
